### PR TITLE
IF: Calculate IF Genesis Block finality digest correctly

### LINKED ISSUE #2348: IF: Genesis block state finality digest correction
When converting a genesis `block_state_legacy` into a `block_state`, the validation tree constructed for the `block_state` includes a single leaf node (the one for the genesis block).

Currently, that leaf node assumes an empty digest for the finality digest of the genesis block. It should instead compute the digest from the constructed `block_header_state`.

### PR DESCRIPTION
**Pull request by heifner**
Resolves #2348

### CODE REVIEW

**Review by greg7mdp — APPROVED** [review]

File: libraries/chain/controller.cpp [hunk 1]
```diff
@@ -763,12 +763,12 @@ struct building_block {
                   // Create the valid structure for validating_bsp if it does not
                   // have one.
                   if (!validating_bsp->valid) {
-                     validating_bsp->valid = bb.parent.new_valid(bhs, action_mroot);
+                     validating_bsp->valid = bb.parent.new_valid(bhs, action_mroot, bb.parent.strong_digest);
```
Comment (linh2931) [inline libraries/chain/controller.cpp:766]: `bb.parent.strong_digest` does not seem correct. It should be the strong digest of the next block state. Here we don't have strong digest yet. Have to compute it by `bhs.compute_finality_digest()`.

**Review by linh2931 — CHANGES_REQUESTED** [review]

File: libraries/chain/controller.cpp (hunk 1 above)
Comment (heifner) [reply libraries/chain/controller.cpp]: Done

**Review by linh2931 — APPROVED** [review]

File: libraries/chain/block_state.cpp [hunk 2]
```diff
@@ -60,45 +60,56 @@ block_state_ptr block_state::create_if_genesis_block(const block_state_legacy& b
...
 
+   assert(bsp.block->contains_header_extension(instant_finality_extension::extension_id())); // required by transition mechanism
+   instant_finality_extension if_ext = bsp.block->extract_header_extension<instant_finality_extension>();
+   assert(if_ext.new_finalizer_policy); // required by transition mechanism
+   result.active_finalizer_policy = std::make_shared<finalizer_policy>(*if_ext.new_finalizer_policy);
+   result.active_proposer_policy = std::make_shared<proposer_policy>();
+   result.active_proposer_policy->active_time = bsp.timestamp();
+   result.active_proposer_policy->proposer_schedule = bsp.active_schedule;
+   result.proposer_policies = {};  // none pending at IF genesis block
+   result.finalizer_policies = {}; // none pending at IF genesis block
```
Comment (greg7mdp) [inline libraries/chain/block_state.cpp:77]: Is this just for documentation?

File: libraries/chain/block_state.cpp [hunk 3]
```diff
@@ -60,45 +60,56 @@ block_state_ptr block_state::create_if_genesis_block(const block_state_legacy& b
...
+
+   // set block_state data ----
+   result.block = bsp.block;
+   result.strong_digest = result.compute_finality_digest(); // all block_header_state data populated in result at this point
+   result.weak_digest = create_weak_digest(result.strong_digest);
+
+   // TODO: https://github.com/AntelopeIO/leap/issues/2057
+   // TODO: Do not aggregate votes on blocks created from block_state_legacy. This can be removed when #2057 complete.
+   result.pending_qc = pending_quorum_certificate{result.active_finalizer_policy->finalizers.size(), result.active_finalizer_policy->threshold, result.active_finalizer_policy->max_weak_sum_before_weak_final()};
+   result.valid_qc = {}; // best qc received from the network inside block extension, empty until first savanna proper IF block
```
Comment (greg7mdp) [inline libraries/chain/block_state.cpp:88]: Is this just for documentation?

File: libraries/chain/block_state.cpp [hunk 4]
```diff
@@ -60,45 +60,56 @@ block_state_ptr block_state::create_if_genesis_block(const block_state_legacy& b
...
+
    // Calculate Merkle tree root in Savanna way so that it is stored in Leaf Node when building block_state.
    auto digests = *bsp.action_receipt_digests_savanna;
    auto action_mroot_svnn = calculate_merkle(std::move(digests));
+
    // built leaf_node and validation_tree
```
Comment (greg7mdp) [inline libraries/chain/block_state.cpp:94]: I think: 

```suggestion
   // build leaf_node and validation_tree
```

File: libraries/chain/block_state.cpp [hunk 5]
```diff
@@ -60,45 +60,56 @@ block_state_ptr block_state::create_if_genesis_block(const block_state_legacy& b
...
-   result.active_proposer_policy->active_time = bsp.timestamp();
-   result.active_proposer_policy->proposer_schedule = bsp.active_schedule;
-   // TODO: https://github.com/AntelopeIO/leap/issues/2057
-   // TODO: Do not aggregate votes on blocks created from block_state_legacy. This can be removed when #2057 complete.
-   result.pending_qc = pending_quorum_certificate{result.active_finalizer_policy->finalizers.size(), result.active_finalizer_policy->threshold, result.active_finalizer_policy->max_weak_sum_before_weak_final()};
    result.validated = bsp.is_valid();
    result.pub_keys_recovered = bsp._pub_keys_recovered;
    result.cached_trxs = bsp._cached_trxs;
    result.action_mroot = action_mroot_svnn;
+   result.base_digest = {}; // calculated on demand in get_finality_data()
```
Comment (greg7mdp) [inline libraries/chain/block_state.cpp:112]: is it useful?

**Review by greg7mdp — APPROVED** [review]

File: libraries/chain/block_state.cpp (hunk 3 above)
Comment (heifner) [reply libraries/chain/block_state.cpp:88]: Yes, Areg wanted all attributes accounted for explicitly.

File: libraries/chain/block_state.cpp (hunk 5 above)
Comment (heifner) [reply libraries/chain/block_state.cpp:112]: Yes, Areg wanted all attributes accounted for explicitly.

**Review by greg7mdp — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: IF
category: INTERNALS
summary: Correct the construction of finality digest when converting genesis block state.
Note:end

**Pull request merged**